HPCC-19048 Unable to refresh locked session

miguelvazq opened the pull request:
When a user locks a session and tries to open a new tab or refreshes the
locked screen they will get sent to an eclwatch without any information
being sent from the server. ESP attempts to redirect but the browser
has cached the content and is stuck in a loop.

Signed-off by: Miguel Vazquez <miguel.vazquez@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 3fbcd271e1cdc012320782d0f0cbbc4eed9f9274
Build: success
ECL Watch:  Rebuilding Site
Rebuild: success

Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |765 |765 |0 |
|test (thor) |679 |679 |0 |
|test (roxie) |792 |792 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  185 sec (00:03:05) |  121 sec (00:02:01) |    7 sec (00:00:07) |   38 sec (00:00:38) | 1121 sec (00:18:41) |   33 sec (00:00:33) | 1527 sec (00:25:27) |

miguelvazq commented:
@GordonSmith Please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: cd4f04674443d01965aba02fa436208cf0676481
Build: success
ECL Watch:  Rebuilding Site
Rebuild: success

Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |765 |765 |0 |
|test (thor) |679 |679 |0 |
|test (roxie) |792 |792 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  179 sec (00:02:59) |  121 sec (00:02:01) |    7 sec (00:00:07) |   41 sec (00:00:41) | 1138 sec (00:18:58) |   32 sec (00:00:32) | 1552 sec (00:25:52) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 5a906ad16e56765e735c8adada8bf9f4906b0479
Build: success
ECL Watch:  Rebuilding Site
Rebuild: success

Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |765 |765 |0 |
|test (thor) |679 |679 |0 |
|test (roxie) |792 |792 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  181 sec (00:03:01) |  119 sec (00:01:59) |    7 sec (00:00:07) |   38 sec (00:00:38) | 1129 sec (00:18:49) |   32 sec (00:00:32) | 1528 sec (00:25:28) |

GordonSmith commented on esp/src/eclwatch/stub.js:
```diff
@@ -154,9 +155,13 @@ define([
-                if (cookie("ESPSessionTimeoutSeconds")) {
+                if (sessionIsActive) {
+                    if (!cookie("ECLWatchUser")) {
+                        dojo.cookie("ECLWatchUser", "true");
```
What is this cookie used for?  My concern is that it can be set and there might be no valid session.

miguelvazq replied:
We check to verify if there is a valid session before setting the cookie. The issue is that ESP will try to redirect to Login.html but, due to the cache of the browser it sends us into a "session-less" ECLWatch repeatedly.

GordonSmith commented on esp/src/eclwatch/stub.js:
```diff
@@ -173,6 +178,8 @@ define([
+                } else if (dojo.cookie("ECLWatchUser")) { //this is due to the hard cache done by the browser
+                    window.location.replace(dojoConfig.urlInfo.basePath + "/Login.html");
                 }
                 stopLoading();
```
Should  the ECLWatchUser cookie be cleared at this point?

miguelvazq replied:
_onLogout() perhaps

miguelvazq replied:
Should be cleared onLogout() I would assume.

richardkchapman commented:
@miguelvazq

GordonSmith reviewed (approved):
Looks ok (not sure I follow the workflow in general though)

Commit:
HPCC-19048 Unable to refresh locked session

When a user locks a session and tries to open a new tab or refreshes the locked screen they will get sent to an eclwatch without any
information being sent from the server. ESP attempts to redirect but the browser
has cached the content and is stuck in a loop.

Signed-off by: Miguel Vazquez <miguel.vazquez@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6013f0267441f5ea55b904c773ed10872473fe14
Build: success
ECL Watch:  Rebuilding Site
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.1.3 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.3: wanted {os:darwin,arch:any} (current: {os:linux,arch:x64})
CPack: Create package

Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
Adding hpcc to group hpcc ...\x1b[32m [   OK    ]  \x1b[0m 
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |758 |758 |0 |
|test (thor) |685 |685 |0 |
|test (roxie) |798 |798 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

[Build error:](https://gist.github.com/579b36ac716fbf228fe501cd08a486f6)
Build log [Build log](https://gist.github.com/HPCCSmoketest/579b36ac716fbf228fe501cd08a486f6#file-result-18-03-21-17-56-28-log)

Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  183 sec (00:03:03) |  120 sec (00:02:00) |    7 sec (00:00:07) |   38 sec (00:00:38) | 1086 sec (00:18:06) |   34 sec (00:00:34) | 1478 sec (00:24:38) |

miguelvazq commented:
Tested on the 190 with my latest commit of clearing ECLWatchUser upon logout to avoid multiple cookie sets.

richardkchapman commented:
@GordonSmith  @AttilaVamos What about the smoketest warnings?

AttilaVamos commented:
@if I remember well, @GordonSmith asked to handle warning as error.

miguelvazq commented:
It has been happening in multiple places such as: https://github.com/hpcc-systems/HPCC-Platform/pull/10981

miguelvazq commented:
This would be more of a question for @GordonSmith. The errors I commonly see and fix are U/I related syntax errors and such.

richardkchapman commented:
I believe they are expected warnings - but if so then smoketest needs to explicitly ignore them. I don't want to become acclimatised to seeing a red cross from the smoketest and yet still proceeding with the merge.

@GordonSmith can you comment?

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 922f2a1f44a1ccb7f3cd6686722e6b2b17fa6323
Build: success
ECL Watch:  Rebuilding Site
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.1.3 (node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.1.3: wanted {os:darwin,arch:any} (current: {os:linux,arch:x64})
CPack: Create package

Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
Adding hpcc to group hpcc ...\x1b[32m [   OK    ]  \x1b[0m 
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

[Build error:](https://gist.github.com/d7d20b6ec20948431261496a53793137)
Build log [Build log](https://gist.github.com/HPCCSmoketest/d7d20b6ec20948431261496a53793137#file-result-18-03-26-17-56-46-log)

Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  181 sec (00:03:01) |  128 sec (00:02:08) |    7 sec (00:00:07) |   23 sec (00:00:23) | 1143 sec (00:19:03) |   17 sec (00:00:17) | 1533 sec (00:25:33) |

GordonSmith commented:
@AttilaVamos those two warnings are expected - it should be a green tick - but also display the warnings?
